pennylane/labs/dla/dense_util.py

```diff
@@ -0,0 +1,456 @@
+
+    Returns:
+        Union[~.Hamiltonian, ~.PauliSentence]: the matrix (matrices) decomposed as a
+        linear combination of Pauli operators, returned either as a :class:`~.Hamiltonian`
+        or :class:`~.PauliSentence` instance.
+
+    .. seealso:: :func:`~.pauli_coefficients`
+
+    **Examples**
+
+    Consider the Hamiltonian :math:`H=\frac{1}{4} X_0 + \frac{2}{5} Z_0 X_1`. We can compute its
+    matrix and get back the Pauli representation via ``pauli_decompose``.
+
+    >>> from pennylane.labs.dla import pauli_decompose
+    >>> H = 1 / 4 * qml.X(0) + 2 / 5 * qml.Z(0) @ qml.X(1)
+    >>> mat = H.matrix()
+    >>> op = pauli_decompose(mat)
+    >>> op
+    0.25 * X(1) + 0.4 * Z(1)
+    >>> type(op)
+    pennylane.ops.op_math.sum.Sum
+
+    We can choose to receive a :class:`~.PauliSentence` instead as output instead, by setting
+    ``pauli=True``:
+
+    >>> op = pauli_decompose(mat, pauli=True)
+    >>> type(op)
+    pennylane.pauli.pauli_arithmetic.PauliSentence
+
+    This function supports batching and will return a list of operations for a batched input:
+
+    >>> ops = [1 / 4 * qml.X(0), 1 / 2 * qml.Z(0) + 1e-7 * qml.Y(0)]
+    >>> batch = np.stack([op.matrix() for op in ops])
+    >>> pauli_decompose(batch)
+    [0.25 * X(0), 1e-07 * Y(0) + 0.5 * Z(0)]
+
+    Small contributions can be removed by specifying the ``tol`` parameter, which defaults
+    to ``1e-10``, accordingly:
+
+    >>> pauli_decompose(batch, tol=1e-6)
+    [0.25 * X(0), 0.5 * Z(0)]
+    """
+    if tol is None:
+        tol = 1e-10
+    coeffs = pauli_coefficients(H)
+    if single_H := qml.math.ndim(coeffs) == 1:
+        coeffs = [coeffs]
+
+    n = int(np.round(np.log2(qml.math.shape(coeffs)[1]))) // 2
+
+    H_ops = []
+    for _coeffs in coeffs:
+        ids = qml.math.where(qml.math.abs(_coeffs) > tol)[0]
+        sentence = PauliSentence({_idx_to_pw(idx, n): c for c, idx in zip(_coeffs[ids], ids)})
+        if pauli:
+            H_ops.append(sentence)
+        else:
+            H_ops.append(sentence.operation())
+
+    if single_H:
+        return H_ops[0]
+    return H_ops
+
+
+def trace_inner_product(
+    A: Union[PauliSentence, Operator, np.ndarray], B: Union[PauliSentence, Operator, np.ndarray]
+):
+    r"""Trace inner product
+
+    Implementation of the trace inner product :math:`\langle A, B \rangle = \text{tr}\left(A^\dagger B\right)/\text{dim}(A)`
+
+    If the inputs are ``np.ndarray``, leading broadcasting axes are supported for either or both
+    inputs.
+
+    Args:
+        A (Union[PauliSentence, Operator, np.ndarray]): First operator
+        B (Union[PauliSentence, Operator, np.ndarray]): Second operator
+
+    Returns:
+        Union[float, np.ndarray]: Result is either a single float or a batch of floats.
+
+    **Example**
+
+    >>> from pennylane.labs.dla import trace_inner_product
+    >>> trace_inner_product(qml.X(0) + qml.Y(0), qml.Y(0) + qml.Z(0))
+    1.0
+
```

Comment:
We should add a batched numpy example I think.

Reply:
✅